File: apps/quickink/ios/QuickInk/Nav/SundialCaptureMenu.swift
/*
 * SundialCaptureMenu.swift
 *
 * Radial capture menu that fans three action buttons out in a 180°
 * arc above the centred ⚡ FAB on the bottom-nav. Tapping the FAB
 * opens the menu; each ray launches a specific capture mode
 * (Document / Business Card / Photo). Replaces the prior
 * tap-for-last-mode + long-press-for-photo idiom with a single
 * explicit choice surface.
 *
 *        Scan
 *      ╱
 *     ●  ← FAB centre (anchor)
 *      ╲
 *        Photo
 *
 *   - Card  → 150° (top-left)
 *   - Scan  →  90° (top, primary)
 *   - Photo →  30° (top-right)
 *
 * Geometry mirrors the design handoff: 110pt radius from the FAB
 * centre, vertical-spring overshoot on open, staggered left → top
 * → right (60ms between rays), simultaneous on close. Dim overlay
 * is the canvas cream at 0.7 + ultra-thin material blur.
 *
 * The menu renders as a full-screen sibling of the NavigationStack
 * so the dim overlay covers the canvas AND the floating nav bar
 * card. Rays sit ABOVE the dim but BELOW the FAB itself — the FAB
 * stays interactive throughout (toggle / close on second tap),
 * matching the prototype's z-stack.
 *
 * Mirror of Android `SundialCaptureMenu.kt`.
 */

import SwiftUI

/// Full-screen overlay that renders the three radial capture
/// buttons. Owns no state — the parent (MainShell) drives
/// `isOpen` and the three select callbacks plumb back into the
/// QuickCapture sheet's `initialMode`.
public struct SundialCaptureMenu: View {

    public let isOpen: Bool
    public let onClose: () -> Void
    public let onSelectDocument: () -> Void
    public let onSelectCard: () -> Void
    public let onSelectPhoto: () -> Void

    public init(
        isOpen: Bool,
        onClose: @escaping () -> Void,
        onSelectDocument: @escaping () -> Void,
        onSelectCard: @escaping () -> Void,
        onSelectPhoto: @escaping () -> Void
    ) {
        self.isOpen = isOpen
        self.onClose = onClose
        self.onSelectDocument = onSelectDocument
        self.onSelectCard = onSelectCard
        self.onSelectPhoto = onSelectPhoto
    }

    // Arc radius from FAB centre to each ray's button centre. 110pt
    // (vs. the spec's 120pt) shaves enough off the leftmost ray that
    // it stays clear of a 320pt iPhone SE's edge by a comfortable
    // margin. Larger devices read the slightly tighter arc as a more
    // confident grouping, not a constraint.
    private let radius: CGFloat = 110

    public var body: some View {
        GeometryReader { proxy in
            // Distance from the screen bottom to the FAB centre. The
            // nav bar uses `padding(.bottom, s6)` = 24pt below its
            // card surface, the card itself is ~64pt tall, and the
            // FAB is lifted -16pt from the card top — so the FAB
            // centre sits roughly safeAreaBottom + 24 + 32 (half
            // bar) + 16 (lift) - 32 (half FAB) = safeAreaBottom + 40.
            // Tuned empirically; matches the FAB drawn in
            // `QuickInkBottomNavBar.swift`.
            let bottomInset = proxy.safeAreaInsets.bottom
            let fabCentreY  = proxy.size.height - (bottomInset + 56)

            ZStack(alignment: .topLeading) {
                // Dim + blur overlay — covers the whole screen
                // including the safe area. Tap dismisses.
                Color.black
                    .opacity(isOpen ? 0.15 : 0)
                    .background(
                        Group {
                            if isOpen {
                                Rectangle()
                                    .fill(QuickInkColors.bg.opacity(0.45))
                                    .background(.ultraThinMaterial)
                            } else {
                                Color.clear
                            }
                        }
                    )
                    .ignoresSafeArea(.all)
                    .contentShape(Rectangle())
                    .onTapGesture { if isOpen { onClose() } }
                    .allowsHitTesting(isOpen)
                    .animation(.easeInOut(duration: 0.28), value: isOpen)

                // Ray anchor — three buttons emanating from the FAB
                // centre point. Each ray's offset is computed from
                // its angle + the shared radius.
                ZStack {
                    ray(
                        label: "Card",
                        systemImage: "person.text.rectangle",
                        angleDeg: 150,
                        openDelay: 0,
                        accessibilityLabel: "Scan business card",
                        action: onSelectCard
                    )
                    ray(
                        label: "Scan",
                        systemImage: "doc.text.viewfinder",
                        angleDeg: 90,
                        openDelay: 60,
                        accessibilityLabel: "Scan document",
                        action: onSelectDocument
                    )
                    ray(
                        label: "Photo",
                        systemImage: "camera",
                        angleDeg: 30,
                        openDelay: 120,
                        accessibilityLabel: "Take photo",
                        action: onSelectPhoto
                    )
                }
                .frame(width: 60, height: 60)
                .position(x: proxy.size.width / 2, y: fabCentreY)
                .allowsHitTesting(isOpen)
            }
        }
    }

    /// Single ray — circular icon button with a label chip beneath.
    /// Closed state: scale 0.4, opacity 0, sitting at the FAB
    /// centre. Open state: scale 1, opacity 1, translated out to its
    /// arc position. Spring + per-ray delay drive the stagger;
    /// reverse animations on close use the same spring with no delay
    /// so all three slingshot back simultaneously.
    @ViewBuilder
    private func ray(
        label: String,
        systemImage: String,
        angleDeg: Double,
        openDelay: Double,
        accessibilityLabel: String,
        action: @escaping () -> Void
    ) -> some View {
        let rad: Double = angleDeg * .pi / 180
        let dx: CGFloat = CGFloat(cos(rad)) * radius
        // CSS Y grows downward; here we negate the sin so positive
        // angles (above horizontal) translate UPWARD on screen.
        let dy: CGFloat = -CGFloat(sin(rad)) * radius

        Button(action: {
            UIImpactFeedbackGenerator(style: .medium).impactOccurred()
            action()
        }) {
            VStack(spacing: 8) {
                ZStack {
                    Circle()
                        .fill(QuickInkColors.surface)
                        .frame(width: 60, height: 60)
                        .shadow(
                            color: QuickInkColors.ink.opacity(0.22),
                            radius: 10,
                            x: 0,
                            y: 5
                        )
                    Image(systemName: systemImage)
                        .font(.system(size: 24, weight: .medium))
                        .foregroundStyle(QuickInkColors.accent)
                }
                Text(label)
                    .font(.system(size: 12, weight: .medium))
                    .foregroundStyle(QuickInkColors.ink)
                    .padding(.horizontal, 10)
                    .padding(.vertical, 4)
                    .background(
                        Capsule().fill(QuickInkColors.surface.opacity(0.95))
                    )
                    .shadow(
                        color: QuickInkColors.ink.opacity(0.10),
                        radius: 4,
                        x: 0,
                        y: 2
                    )
                    .fixedSize()
            }
        }
        .buttonStyle(.plain)
        .scaleEffect(isOpen ? 1.0 : 0.4, anchor: .top)
        .offset(x: isOpen ? dx : 0, y: isOpen ? dy : 0)
        .opacity(isOpen ? 1.0 : 0)
        .animation(
            .interpolatingSpring(stiffness: 200, damping: 16)
                .delay(isOpen ? openDelay / 1000 : 0),
            value: isOpen
        )
        .accessibilityLabel(Text(accessibilityLabel))
    }
}

File: apps/quickink/android/app/src/main/java/app/quickink/mobile/features/nav/SundialCaptureMenu.kt
/*
 * SundialCaptureMenu.kt
 *
 * Radial capture menu that fans three action buttons out in a 180°
 * arc above the centred ⚡ FAB on the bottom-nav. Tapping the FAB
 * opens the menu; each ray launches a specific capture mode
 * (Document / Business Card / Photo). Replaces the prior
 * tap-for-last-mode + long-press-for-photo idiom with a single
 * explicit choice surface.
 *
 *        Video
 *      ╱
 *     ●  ← FAB centre (anchor)
 *      ╲
 *        Photo
 *
 *   - Scan  → 150° (top-left)
 *   - Video →  90° (top, primary)
 *   - Photo →  30° (top-right)
 *
 * Video and Photo both open the [PhotoCaptureSurface], which
 * supports both still capture (tap shutter) and hold-to-record
 * video on the same camera session. The two rays exist so users
 * who think "I want a video" don't have to know that internally
 * we share a surface — they pick the verb that matches their
 * intent and the surface adapts.
 *
 * Geometry mirrors the design handoff: 110dp radius from the FAB
 * centre, vertical-spring overshoot on open, staggered left → top
 * → right (60ms between rays), simultaneous on close. Dim overlay
 * is the canvas cream at 0.7 alpha.
 *
 * The menu renders as a full-screen sibling of the NavHost so the
 * dim layer covers the canvas AND the floating nav bar card. Rays
 * sit ABOVE the dim but BELOW the FAB itself — the FAB stays
 * interactive throughout (toggle / close on second tap), matching
 * the prototype's z-stack.
 *
 * Mirror of iOS `SundialCaptureMenu.swift`.
 */

package app.quickink.mobile.features.nav

import androidx.compose.animation.core.Spring
import androidx.compose.animation.core.animateFloatAsState
import androidx.compose.animation.core.spring
import androidx.compose.animation.core.tween
import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.interaction.MutableInteractionSource
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.offset
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.outlined.CameraAlt
import androidx.compose.material.icons.outlined.DocumentScanner
import androidx.compose.material.icons.outlined.Videocam
import androidx.compose.material3.Icon
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.alpha
import androidx.compose.ui.draw.scale
import androidx.compose.ui.draw.shadow
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.TransformOrigin
import androidx.compose.ui.graphics.graphicsLayer
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.platform.LocalDensity
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import app.quickink.mobile.ui.theme.LocalQuickInkColors
import app.quickink.mobile.ui.theme.LocalQuickInkTypography
import kotlin.math.cos
import kotlin.math.sin

/**
 * Full-screen overlay that renders the three radial capture
 * buttons. Owns no state — the parent (MainShell) drives
 * [isOpen] and the three select callbacks plumb back into the
 * QuickCapture sheet's `initialMode`.
 */
@Composable
fun SundialCaptureMenu(
    isOpen: Boolean,
    onClose: () -> Unit,
    onSelectScan: () -> Unit,
    onSelectVideo: () -> Unit,
    onSelectPhoto: () -> Unit,
    modifier: Modifier = Modifier,
) {
    val colors = LocalQuickInkColors.current

    // The dim overlay animates between fully transparent (closed)
    // and the cream-tinted backdrop (open). Tap on it dismisses;
    // disabled while closed so taps fall through to the canvas.
    val overlayAlpha by animateFloatAsState(
        targetValue = if (isOpen) 1f else 0f,
        animationSpec = tween(durationMillis = 280),
        label = "sundial-overlay-alpha",
    )

    val interactionSource = remember { MutableInteractionSource() }
    Box(
        modifier = modifier.fillMaxSize(),
        contentAlignment = Alignment.BottomCenter,
    ) {
        // Dim layer — `colors.bg` (canvas cream) at 0.72 alpha so
        // the bar + canvas read through faintly without being
        // obtrusive. Drawn first so it sits BEHIND the rays.
        if (overlayAlpha > 0f) {
            Box(
                modifier = Modifier
                    .fillMaxSize()
                    .alpha(overlayAlpha)
                    .background(colors.bg.copy(alpha = 0.72f))
                    .clickable(
                        interactionSource = interactionSource,
                        indication        = null,
                        enabled           = isOpen,
                        onClick           = onClose,
                    ),
            )
        }

        // Ray anchor — positioned at the FAB centre. The nav bar
        // sits at safeArea + s6 (24dp) from the bottom; its card
        // is ~64dp tall; the FAB inside is lifted -16dp from the
        // card top — so the FAB centre is approximately
        // safeArea + 24 + 32 (half bar) - 16 (lift) + 0 = +40dp
        // above the bar bottom. Mirror that with padding here so
        // the rays emanate from the bolt.
        Box(
            modifier = Modifier
                .padding(bottom = 56.dp)
                .size(60.dp),
            contentAlignment = Alignment.Center,
        ) {
            Ray(
                label              = "Scan",
                icon               = Icons.Outlined.DocumentScanner,
                angleDeg           = 150.0,
                openDelayMs        = 0,
                accessibilityLabel = "Scan document",
                isOpen             = isOpen,
                onClick            = onSelectScan,
            )
            Ray(
                label              = "Video",
                icon               = Icons.Outlined.Videocam,
                angleDeg           = 90.0,
                openDelayMs        = 60,
                accessibilityLabel = "Record video",
                isOpen             = isOpen,
                onClick            = onSelectVideo,
            )
            Ray(
                label              = "Photo",
                icon               = Icons.Outlined.CameraAlt,
                angleDeg           = 30.0,
                openDelayMs        = 120,
                accessibilityLabel = "Take photo",
                isOpen             = isOpen,
                onClick            = onSelectPhoto,
            )
        }
    }
}

/**
 * Arc radius from FAB centre to each ray's button centre. 110dp
 * (vs. the spec's 120) leaves the leftmost ray clear of the
 * 320dp edge of small-screen devices.
 */
private val RAY_RADIUS = 110.dp

@Composable
private fun Ray(
    label: String,
    icon: ImageVector,
    angleDeg: Double,
    openDelayMs: Int,
    accessibilityLabel: String,
    isOpen: Boolean,
    onClick: () -> Unit,
) {
    val colors  = LocalQuickInkColors.current
    val type    = LocalQuickInkTypography.current
    val density = LocalDensity.current

    val rad = angleDeg * Math.PI / 180.0
    val dx  = (cos(rad) * RAY_RADIUS.value).dp
    // Compose Y grows downward — negate so positive angles
    // (above horizontal) translate UPWARD on screen.
    val dy  = (-sin(rad) * RAY_RADIUS.value).dp

    val animSpec = spring<Float>(
        dampingRatio = 0.55f,
        stiffness    = Spring.StiffnessMediumLow,
    )

    // animateFloatAsState lets us drive `scale`, `opacity`, and
    // translation through the same spring. Delay only applied
    // when opening (close fires all three rays together for a
    // snappy dismiss).
    val progress by animateFloatAsState(
        targetValue   = if (isOpen) 1f else 0f,
        animationSpec = if (isOpen) {
            spring(dampingRatio = 0.55f, stiffness = Spring.StiffnessMediumLow)
        } else {
            tween(durationMillis = 240)
        },
        label = "sundial-ray-progress-$label",
    )

    val translatedX = with(density) { (dx.toPx()) * progress }
    val translatedY = with(density) { (dy.toPx()) * progress }
    val scale       = 0.4f + (0.6f * progress)
    val rayAlpha    = progress

    val interactionSource = remember { MutableInteractionSource() }
    Column(
        horizontalAlignment   = Alignment.CenterHorizontally,
        verticalArrangement   = Arrangement.spacedBy(8.dp),
        modifier = Modifier
            .graphicsLayer {
                translationX    = translatedX
                translationY    = translatedY
                scaleX          = scale
                scaleY          = scale
                this.alpha      = rayAlpha
                transformOrigin = TransformOrigin(0.5f, 0f)
            }
            .semantics { contentDescription = accessibilityLabel }
            .clickable(
                interactionSource = interactionSource,
                indication        = null,
                enabled           = isOpen,
                onClick           = onClick,
            ),
    ) {
        Box(
            modifier = Modifier
                .size(60.dp)
                .shadow(elevation = 10.dp, shape = CircleShape, clip = false)
                .background(colors.surface, CircleShape),
            contentAlignment = Alignment.Center,
        ) {
            Icon(
                imageVector        = icon,
                contentDescription = null,
                tint               = colors.accent,
                modifier           = Modifier.size(26.dp),
            )
        }
        Box(
            modifier = Modifier
                .shadow(elevation = 4.dp, shape = RoundedCornerShape(11.dp))
                .background(
                    color = colors.surface.copy(alpha = 0.95f),
                    shape = RoundedCornerShape(11.dp),
                )
                .padding(horizontal = 10.dp, vertical = 4.dp),
        ) {
            Text(
                text     = label,
                style    = type.label.copy(fontSize = 12.sp),
                color    = colors.ink,
                maxLines = 1,
                softWrap = false,
            )
        }
    }
}
